### PR TITLE
Node.js SDK - CLI Downloader

### PR DESCRIPTION
**Pull request by myty**
Additional changes identified from #4390 

* separate cli-downloader functionality from Bin class and move it into its own class, CliDownloader
* updates tests

### CODE REVIEW

File: sdk/nodejs/provisioning/cli-downloader.ts [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+}
+
+export class CliDownloader {
+  private readonly cliVersion: string
+  private readonly archive?: CliDownloaderOptions["archive"]
+  private readonly executableFilename: (name: string) => string
+
+  constructor(options: CliDownloaderOptions) {
+    if (!options.cliVersion) {
+      throw new Error("cliVersion is not set")
```
Comment (myty) [inline sdk/nodejs/provisioning/cli-downloader.ts:38]: @TomChv I'm looking through the different error class in the `error.ts`, but unsure which would be best to use.  I was thinking of this one, `InitEngineSessionBinaryError`, but maybe we need a new one. Any recommendations?

File: sdk/nodejs/provisioning/cli-downloader.ts [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+  static async Download(options: CliDownloaderOptions): Promise<string> {
+    const cliDownloader =
+      os.platform() === "win32"
+        ? new WindowsCliDownloader(options)
+        : new CliDownloader(options)
+
+    return await cliDownloader.Download()
+  }
+
+  async Download(): Promise<string> {
```
Comment (myty) [inline sdk/nodejs/provisioning/cli-downloader.ts:56]: @TomChv Just referencing your comment from the previous PR, I tend to like pascalCase method names in javascript files, but I was following, what appeared to be, a convention of public methods being PascalCase vs camelCase for private methods. Maybe that was an incorrect assumption on my part?

File: sdk/nodejs/provisioning/cli-downloader.ts (hunk 1 above)
Comment (slumbering) [reply sdk/nodejs/provisioning/cli-downloader.ts:38]: Hey @myty ! Using `InitEngineSessionBinaryError` make total sense IMO

File: sdk/nodejs/provisioning/cli-downloader.ts (hunk 2 above)
Comment (slumbering) [reply sdk/nodejs/provisioning/cli-downloader.ts:56]: AFAIK, Class methods are using lowerCamelCase by convention.
- https://google.github.io/styleguide/jsguide.html#naming-method-names
- https://github.com/airbnb/javascript#naming--camelCase

We should stay consistent in capitalization with the rest of our codebase. 
Formatting is still pretty subjective though ;)
- https://github.com/ryanmcdermott/clean-code-javascript#formatting

Comment (myty) [reply sdk/nodejs/provisioning/cli-downloader.ts:56]: I'm totally down for that.

For future reference, the other classes in provisioning folder will need updated for consistency.  For instance, https://github.com/dagger/dagger/blob/main/sdk/nodejs/provisioning/bin.ts

Comment (slumbering) [reply sdk/nodejs/provisioning/cli-downloader.ts:56]: 100% agreed

File: sdk/nodejs/provisioning/cli-downloader/cli-downloader.ts [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+  }
+
+  private cliArchiveName(): string {
+    const architecture = this.normalizedArch()
+
+    if (this.archive?.name != null) {
+      return this.archive.name(architecture)
+    }
+
+    return `dagger_v${this.cliVersion}_${os.platform()}_${architecture}.tar.gz`
```
Comment (gerhard) [inline sdk/nodejs/provisioning/cli-downloader/cli-downloader.ts:158]: Extract `os.platform()` in a constant above, same as `architecture`

**Review by gerhard — COMMENTED** [review]
Thanks for this PR! We did a quick review with @slumbering & @dolanor. There are a few more things that they will be adding before we can move this along. Our intention is to have this merged by the end of the week.

The biggest thing that we are missing right now is an integration test that gives us confidence in this change. @dolanor @slumbering will follow-up on this.

Thanks @myty 👍

**Review by dolanor — CHANGES_REQUESTED** [review]
Hi @myty,
Thanks for your contribution.

I have 2 questions/suggestions.
I'm also for the integration tests to make sure nothing is broken in between as the change is big and change the logic quite a bit :)

File: sdk/nodejs/connect.ts [hunk 4]
```diff
@@ -86,3 +87,22 @@ export async function connect(
...
+
+/**
+ * Prefer _EXPERIMENTAL_DAGGER_CLI_BIN
+ * with fallback behavior of downloading the CLI and using that as the bin.
+ * @returns the path to the cli executable
+ *
+ */
+async function getCliBinPath() {
+  const cliBinEnvPath = process.env["_EXPERIMENTAL_DAGGER_CLI_BIN"]
+  if (cliBinEnvPath) {
```
[the remark this answers is not in the document]
Comment (myty) [reply sdk/nodejs/connect.ts:99]: That's a good point.  Throw if the path doesn't exist?

[the remark this answers is not in the document]
Comment (dolanor) [reply sdk/nodejs/connect.ts:101]: That seems the best and decoupled thing to do. Thanks.

File: sdk/nodejs/connect.ts [hunk 5]
```diff
@@ -86,3 +89,28 @@ export async function connect(
...
+    if (!fs.existsSync(cliBinEnvPath)) {
+      throw new InitEngineSessionBinaryError(
+        "Dagge CLI path was provided but the file path does nor exist."
```
Comment (dolanor) [inline sdk/nodejs/connect.ts:104]:
```suggestion
        "Dagger CLI path was provided but the file path does not exist."
```

**Comment by dolanor** [comment]
Hi @myty , 
We didn't manage to finish last week.
We spent time with @slumbering today, and I think I found some flaws in the logic and as well in our tests. So we won't merge until we make sure our tests work, and that the implementation works. I hope I can dedicate time this week.

**Pull request closed without merging**